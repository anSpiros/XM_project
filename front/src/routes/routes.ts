export const ROOT_ROUTE: string = "/";
export const FORM_ROUTE: string = "/form";
export const RESULTS_ROUTE: string = "/results";
