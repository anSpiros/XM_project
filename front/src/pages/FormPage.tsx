import React from "react";
import Form from "../components/Form";
import { addResults } from "../state/actions/results";
import { RESULTS_ROUTE } from "../routes/routes";
import { useHistory } from "react-router-dom";
import { useDispatch } from "react-redux";

function FormPage() {
  let history = useHistory();
  const dispatch = useDispatch();
  const onSubmit = (data: any) => {
    dispatch(addResults(data));
    history.push(RESULTS_ROUTE);
  };
  return (
    <div className="row">
      <div className="col-12">
        <Form onSubmit={onSubmit} />
      </div>
    </div>
  );
}

export default FormPage;
