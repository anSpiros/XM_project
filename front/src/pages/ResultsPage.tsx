import React, { useEffect, useState } from "react";
import { DisplayDataInterface } from "../interfaces/display-data.interface";
import _ from "lodash";
import "../scss/Results.scss";
import { Link } from "react-router-dom";
import { FORM_ROUTE } from "../routes/routes";
import { useSelector } from "react-redux";
import { ResultState } from "../state/types/results/types";
import { DisplayData } from "../interfaces/results.interface";
import Results from "../components/Results";

function ResultsPage() {
  const results = useSelector((state: ResultState) => state.results);
  const [chartData, setChartData] = useState({});
  const [companyDescription, setCompanyDescription] = useState("");
  const [displayData, setDisplayData] = useState<DisplayData>({
    body: [],
    header: [],
  });
  const chart = () => {
    setChartData({
      labels: displayData.body.map((data) => {
        return data.date;
      }),
      datasets: [
        {
          label: "open",
          data: displayData.body.map((data) => {
            return data.open;
          }),
          backgroundColor: "rgba(75, 192, 192, 0.6)",
          borderWidth: 4,
        },
        {
          label: "close",
          data: displayData.body.map((data) => {
            return data.close;
          }),
          backgroundColor: "rgba(75, 192, 192, 0.6)",
          borderWidth: 4,
        },
      ],
    });
  };
  const prepareDataToDisplay = (data: any) => {
    const body = data.map((data: any) => {
      const newData: DisplayDataInterface = {
        date: data[0],
        open: data[1],
        high: data[2],
        low: data[3],
        close: data[4],
        volume: data[5],
      };
      return newData;
    });
    setDisplayData({
      body: body,
      header: [
        {
          column: "Date",
        },
        {
          column: "Open",
        },
        {
          column: "High",
        },
        {
          column: "Low",
        },
        {
          column: "Close",
        },
        {
          column: "Volume",
        },
      ],
    });
  };

  const shouldRenderTable = () => {
    return !_.isEmpty(displayData.header) && !_.isEmpty(displayData.body);
  };

  useEffect(() => {
    if (!_.isEmpty(results)) {
      setCompanyDescription(results.dataset.name);
      prepareDataToDisplay(results.dataset.data);
    }
  }, [results]);

  useEffect(() => {
    chart();
  }, [displayData]);

  const chartOptions = {
    responsive: true,
    scales: {
      yAxes: [
        {
          ticks: {
            beginAtZero: true,
            stacked: true,
          },
          gridLines: {
            display: false,
          },
        },
      ],
      xAxes: [
        {
          ticks: {
            stacked: true,
          },
          gridLines: {
            display: false,
          },
        },
      ],
    },
  };
  return (
    <>
      {!_.isEmpty(results) ? (
        <>
          {results.dataset.data.length > 0 ? (
            <>
              {shouldRenderTable() ? (
                <Results
                  chartData={chartData}
                  chartOptions={chartOptions}
                  companyDescription={companyDescription}
                  displayDataForTable={displayData}
                />
              ) : (
                <div className="results-loading">
                  <h2>Loading...</h2>
                </div>
              )}
            </>
          ) : (
            <div className="results-no-data">
              <h2>
                No data to show. Go back to form to change your preferences.
              </h2>
              <Link to={FORM_ROUTE}>Go to form</Link>
            </div>
          )}
        </>
      ) : (
        <div className="results no-data">
          <h2>You must go to form and add your preferences first.</h2>
          <Link to={FORM_ROUTE}>Go to form</Link>
        </div>
      )}
    </>
  );
}

export default ResultsPage;
