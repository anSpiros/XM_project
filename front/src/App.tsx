import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { FORM_ROUTE, RESULTS_ROUTE, ROOT_ROUTE } from "./routes/routes";
import FormPage from "./pages/FormPage";
import ResultsPage from "./pages/ResultsPage";
import Header from "./components/Header";
import HomePage from "./pages/HomePage";

function App() {
  return (
    <Router>
      <div className="App">
        <Header />
        <div className="container mt-5 d-flex justify-content-center">
          <Switch>
            <Route exact path={ROOT_ROUTE}>
              <HomePage />
            </Route>
            <Route path={FORM_ROUTE}>
              <FormPage />
            </Route>
            <Route path={RESULTS_ROUTE}>
              <ResultsPage />
            </Route>
          </Switch>
        </div>
      </div>
    </Router>
  );
}

export default App;
