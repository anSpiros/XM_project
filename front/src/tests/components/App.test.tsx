import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import App from "../../App";
import { Router } from "react-router-dom";
import { Provider } from "react-redux";
import { createStore } from "redux";
import allReducers from "../../state/reducers";
import { createMemoryHistory } from "history";
import { FORM_ROUTE, ROOT_ROUTE } from "../../routes/routes";
import FormPage from "../../pages/FormPage";
import HomePage from "../../pages/HomePage";
import ResultsPage from "../../pages/ResultsPage";

const store = createStore(allReducers);

describe("App component", () => {
  afterEach(cleanup);

  test("renders learn react link", () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router history={history}>
          <App />
        </Router>
      </Provider>
    );
    const linkElement = screen.getByText(/Welcome/i);
    expect(linkElement).toBeInTheDocument();
  });

  test("landing Form page", () => {
    const history = createMemoryHistory();
    history.push(FORM_ROUTE);
    render(
      <Provider store={store}>
        <Router history={history}>
          <FormPage />
        </Router>
      </Provider>
    );

    expect(screen.getByText(/Email/i)).toBeInTheDocument();
  });

  test("Landing Home page", () => {
    const history = createMemoryHistory();
    history.push(ROOT_ROUTE);
    render(<HomePage />);
    expect(screen.getByText(/Welcome/i)).toBeInTheDocument();
  });

  test("Landing Results page", () => {
    const history = createMemoryHistory();
    history.push(ROOT_ROUTE);
    render(
      <Provider store={store}>
        <Router history={history}>
          <ResultsPage />
        </Router>
      </Provider>
    );
    expect(
      screen.getByText(/You must go to form and add your preferences first./i)
    ).toBeInTheDocument();
  });

  test("should render correctly", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store}>
        <Router history={history}>
          <App />
        </Router>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
