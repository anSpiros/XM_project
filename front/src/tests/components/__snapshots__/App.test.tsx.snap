// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="container-fluid header"
    >
      <div
        class="row"
      >
        <div
          class="col-12"
        >
          <ul
            class="list-unstyled header-list mb-0"
          >
            <li>
              <a
                aria-current="page"
                class="active"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                href="/form"
              >
                Form
              </a>
            </li>
            <li>
              <a
                href="/results"
              >
                Results
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="container mt-5 d-flex justify-content-center"
    >
      <div>
        <h1>
          Welcome
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
